Sources/PredicateBuilder/PredicateBuilder.swift

```diff
@@ -0,0 +1,184 @@
+}
+
+// MARK: Predicate Builder
+
+public class PredicateBuilder: PredicateComposing {
+
+  private let type: PredicateOperation.LogicalType
+  private var predicates: [PredicateProvider] = []
+
+  public var predicate: NSPredicate {
+    CompoundPredicate(
+      type: type,
+      subpredicates: predicates
+    ).predicate
+  }
+
+  public init(type: PredicateOperation.LogicalType = .and) {
+    self.type = type
+  }
+
+  public func predicate(key: String,
+                        value: CVarArg,
+                        comparison: PredicateOperation.Comparison,
+                        aggregate: PredicateOperation.Aggregate = .skip,
+                        options: [PredicateOperation.Option] = []) -> PredicateComposing {
+    predicates.append(Predicate(key: key, value: value, comparison: comparison, aggregate: aggregate, options: options))
+    return self
+  }
+
+  public func predicate(key: String,
+                        value: CVarArg,
+                        comparison: PredicateOperation.Comparison,
+                        options: [PredicateOperation.Option]) -> PredicateComposing {
+    predicate(key: key, value: value, comparison: comparison, aggregate: .skip, options: options)
+  }
+
+  public func predicate(key: String,
+                        value: CVarArg,
+                        comparison: PredicateOperation.Comparison) -> PredicateComposing {
+    predicate(key: key, value: value, comparison: comparison, aggregate: .skip, options: [])
+  }
+
+  public func compoundPredicate(type: PredicateOperation.LogicalType,
+                                handler: (PredicateComposing) -> PredicateComposing) -> PredicateComposing {
+    predicates.append(handler(PredicateBuilder(type: type)))
+    return self
+  }
+}
+
+
+// MARK: Predicate Operation
+
+public enum PredicateOperation {
+  public enum Aggregate: String {
+    case skip = ""
+    case any = "ANY"
+    case all = "ALL"
+    case notAny = "NOT ANY"
+  }
+
+  public enum Comparison: String {
+    case equal = "=="
+    case greaterEqual = ">="
+    case greater = ">"
+    case lessEqual = "<="
+    case less = "<"
+    case notEqual = "!="
+    case between = "BETWEEN"
+    case `in` = "IN"
+    case beginsWith = "BEGINSWITH"
+    case contains = "CONTAINS"
+    case endsWith = "ENDSWITH"
+    case like = "LIKE"
+    case matches = "MATCHES"
+  }
+
+  public enum Option: String {
+    case caseInsensitive = "c"
+    case diacriticInsensitive = "d"
+    case normalized = "n"
+    case localeSensitive = "l"
+  }
+
+  public enum LogicalType {
+    case and
+    case or
+    case not
+
+    var type: NSCompoundPredicate.LogicalType {
+      switch self {
+      case .and: return .and
+      case .or: return .or
+      case .not: return .not
+      }
+    }
+  }
+}
+
+// MARK: Predicate
+
+struct Predicate: PredicateProvider {
+
+  private let key: String
+  private let value: CVarArg
+  private var options: [PredicateOperation.Option]
```

Comment:
Ok. So just Predicate instead of PredicateBuilder. It's more declarative.
So I'll be expecting: 
let p = Predicate(...)
let q = Predicate(...)
let r = p.by(.and, with: q), and then the operator p && q on predicates [extasy] ;d

(just a predicate tree)